Add Python fluent & client batched-logging APIs & unit tests

smurching opened the pull request:
Adds fluent & client APIs for batched logging, plus simple unit tests (the stores are tested more heavily in #950).

A follow-up (#956) will add validation logic to the OSS client and backing stores (e.g. checking that a reasonable number of metrics/params/tags are being logged in the client API before attempting to write them to the backing store).

smurching committed:
Add local batch-logging functionality

smurching committed:
Copy over proto changes

smurching committed:
Fix some linter errors & add / fix some more DB tests

smurching committed:
Reemove fluent API & add FileStore tests

smurching committed:
Remove change to duplicate-timestamp behavior in SQL log_metric (to be discussed offline first)

smurching committed:
Remove unnecessary line deletion

smurching committed:
Remove validation

smurching committed:
Remove use of validation helpers

smurching committed:
Remove unneeded exception-error-code-change

smurching committed:
Remove validations for now

smurching committed:
Fix test + improve abstract store signature of log_batch (use run_id instead of run_uuid)

smurching committed:
Merge branch 'batched-logging-local' into batched-logging-client-fluent

smurching committed:
Use run_id instead of run_uuid in client API

smurching committed:
Add new store tests, remove TODO

smurching committed:
Merge branch 'batched-logging-local' into batched-logging-client-fluent

smurching committed:
Undo all changes specific to this PR

smurching committed:
Revert "Undo all changes specific to this PR"

This reverts commit 7a3e62916970c28e207a7be6d6831bb2db0ec175.

smurching committed:
Merge branch 'master' into batched-logging-client-fluent

smurching committed:
Add more tests for fluent API + logic for entity equality & tests

smurching committed:
Test inequality case too

smurching committed:
Undo entity equality changes

smurching committed:
Fix tests w/o entity equality changes

smurching commented on tests/tracking/test_fluent.py:
```diff
@@ -397,3 +398,61 @@ def test_start_run_existing_run_deleted(empty_active_run_stack):
+    (mlflow.set_tags, {"tags": {"d": "e"}},
+     {"run_id": "my-uuid", "metrics": [], "params": [], "tags": [RunTag("d", "e")]}),
+])
+def test_log_batch_apis_delegate_to_store(
```
This test and `test_log_batch_client_apis_delegate_to_store` violate the "test the interface not the implementation" principle, but the alternatives would be having no such tests or duplicating all the `store.log_batch` tests for the fluent/client API, which seems excessive (better to just know that the fluent/client APIs correctly call `store.log_batch` and trusting the stores' unit test coverage, especially since it seems unlikely for the fluent/client APIs to undergo a major refactor that'd require deleting/majorly reworking these tests).

dbczumar replied:
These tests don't seem to cover any function/input cases that aren't already covered by `test_log_batch`, `test_log_metrics`, `test_set_tags`, or `test_log_params`. We also don't verify that existing fluent/client `log_...()` calls invoke appropriate methods in the `AbstractStore`. Is this test necessary?

smurching replied:
Yeah it's a good point that we kinda cover this behavior w/ the other tests (like `test_log_metrics` etc) anyways, happy to delete it :P (and the client API test too)

smurching committed:
Fix lint

dbczumar commented on mlflow/tracking/fluent.py:
```diff
@@ -212,6 +212,40 @@ def log_metric(key, value):
+    """
+    run_id = _get_or_start_run().info.run_uuid
+    timestamp = int(time.time())
+    metrics_arr = [Metric(key, value, timestamp) for key, value in metrics.items()]
```
Looks like we don't perform the same metric value type validation that we do for `log_metric()`. As a result, the following code successfully writes a metric with the value `"dog"` to the file store. However, this corrupts the file store's run representation, and all subsequent calls to `get_run()` with the specified `run_id` fail.

```
with mlflow.start_run():
    run_id = mlflow.active_run().info.run_uuid
    # Writes the metric string value and fails
    mlflow.log_metrics(metrics={"cat": "dog"})
```

```
from mlflow.tracking import MlflowClient
client = MlflowClient()
# Fails because the run has been corrupted by a metric value of type string.
client.get_run(run_id)
```

dbczumar replied:
I still think that it's ideal to perform this kind of validation in the store itself so that every client invocation doesn't have to repeat the same validation logic @mparkhe

smurching replied:
Ah good catch

dbczumar commented on mlflow/tracking/fluent.py:
```diff
@@ -212,6 +212,40 @@ def log_metric(key, value):
+def log_params(params):
+    """
+    Log a batch of params for the current run, starting a run if no runs are active.
+    :param params: Dictionary of param_name: String -> value: String
```
We should mention that non-string values will be stringified, like we do in the documentation for `log_param()`.

dbczumar commented on mlflow/tracking/fluent.py:
```diff
@@ -212,6 +212,40 @@ def log_metric(key, value):
+def set_tags(tags):
+    """
+    Log a batch of tags for the current run, starting a run if no runs are active.
+    :param tags: Dictionary of tag_name: String -> value: String
```
We should mention that non-string values will be stringified, like we do in the documentation for log_param().

dbczumar commented on tests/tracking/test_fluent.py:
```diff
@@ -397,3 +398,61 @@ def test_start_run_existing_run_deleted(empty_active_run_stack):
+               [dict(t) for t in expected_store_kwargs["tags"]]
+
+
+def test_log_batch_client_apis_delegate_to_store(
```
Same question about the necessity of including this test. This also doesn't seem to test fluent API behavior, so I wonder if it belongs in the fluent API test suite.

dbczumar reviewed (changes requested):
@smurching Looks good in general! Left a few comments about metric value validation, tests, documentation

smurching committed:
Address review comments

smurching commented:
Thanks @dbczumar (and @mparkhe in https://github.com/mlflow/mlflow/pull/956) for the feedback - I think I've addressed feedback on this PR, besides the point about doing metric-value validation at the store level - it makes sense to me to do so, I'll add it in #956 if that's alright

smurching commented:
To clarify, I did add metric-value validation to the client API

acroz commented:
What's the expected behaviour when `log_batch()` gets multiple tags or params with the same key, or multiple metrics with the same (key, timestamp) pair? Should it be considered a failure, and if not, should e.g. the last one accepted?

It might be good to include a comment in `AbstractStore.log_batch()` to indicate the expected behaviour for store implementations.

acroz commented:
Also, since the single-value log_metric, log_param and set_tag methods are no longer abstract methods in the `AbstractStore` (and hence are not enforced to be implemented in implementations of the `AbstractStore`, should they have a default implementation that proxies calls to `log_batch`? 

Sorry that this is only tangentially related to this PR, but since you were working on these changes it seemed a relevant place to post this.

smurching committed:
Merge branch 'master' into batched-logging-client-fluent

dbczumar commented on mlflow/tracking/client.py:
```diff
@@ -131,8 +131,8 @@ def log_metric(self, run_id, key, value, timestamp=None):
         """
-        _validate_metric_name(key)
         timestamp = timestamp if timestamp is not None else int(time.time())
+        _validate_metric(key, value, timestamp)
```
Should we call this in `log_batch` as well?

smurching replied:
Oops yeah lol...

dbczumar commented on mlflow/utils/validation.py:
```diff
@@ -33,26 +36,48 @@ def path_not_unique(name):
+            INVALID_PARAMETER_VALUE)
+
+    if not isinstance(timestamp, numbers.Number) or timestamp > np.finfo(np.int64).max \
+            or timestamp < np.finfo(np.int64).min:
```
Nit: Should we instead check that timestamp is non-negative?

dbczumar commented on tests/tracking/test_fluent.py:
```diff
@@ -7,7 +7,7 @@
 from six.moves import reload_module as reload
```
Revert changes to this file

dbczumar reviewed (approved):
LGTM with a couple of minor comments to address

smurching commented:
@acroz good questions, sorry for the delay - it definitely makes sense to clarify the behavior, I'll make a PR to do that.

> What's the expected behaviour when log_batch() gets multiple tags or params with the same key, or multiple metrics with the same (key, timestamp) pair? Should it be considered a failure, and if not, should e.g. the last one accepted? It might be good to include a comment in AbstractStore.log_batch() to indicate the expected behaviour for store implementations.

In general we expect overwriting a param with a new value to fail (e.g. `log_param("a", "b")` followed by `log_param("a", "c")` should fail, but calling `log_param("a", "b")` twice is fine). Thus logging a param twice with different values in the same batched-logging API request should fail. The API doesn't make strong guarantees around whether partial data may be persisted, so some data may be written (e.g. other metrics/params/tags) when such a request is received. 

Tags on the other hand can be overwritten - when logging multiple tags with the same key/different values in the same request, we'd expect the last tag value to win out & be persisted.

For metrics, we do want to allow multiple metrics to be written with the same (metric_key, metric_value, timestamp), so such requests should succeed (note that the current SQLAlchemyStore implementation disallows this, but we will likely change that behavior).

> Also, since the single-value log_metric, log_param and set_tag methods are no longer abstract methods in the AbstractStore (and hence are not enforced to be implemented in implementations of the AbstractStore, should they have a default implementation that proxies calls to log_batch?

That seems reasonable to me, thanks for the heads up on that. Happy to make a follow-up (or review a follow-up) that makes this change (adding a default implementation that proxies calls to log_batch)

acroz commented:
@smurching Thanks for the clarifications!

> For metrics, we do want to allow multiple metrics to be written with the same (metric_key, metric_value, timestamp), so such requests should succeed.

In this case do you expect that logging multiple metrics with a given (key, timestamp) pair would result in an overwrite of the existing value, or that there would be metrics with duplicate (key, timestamp).

For the sake of clarity, if I do (pseudo-code):

```
log_metric(key="key", time=1234567, value=1.0)
log_metric(key="key", time=1234567, value=5.0)
```

Then on querying for metrics do I expect to get:

```json
[
    {"key": "key", "time": 1234567, "value": 5.0}
]
```

or:

 ```json
[
    {"key": "key", "time": 1234567, "value": 1.0},
    {"key": "key", "time": 1234567, "value": 5.0}
]
```

> Happy to make a follow-up (or review a follow-up) that makes this change (adding a default implementation that proxies calls to log_batch)

I'm happy to put this in (it's a small PR), however it's since occurred to me that if the store is only used by the client, maybe these single {tag,metric,param} methods should just be removed completely?

smurching committed:
Address review comments / fix tests after merging master

smurching commented:
@acroz for the metrics question, we should expect to see the latter:

```
[
   {"key": "key", "time": 1234567, "value": 1.0},
   {"key": "key", "time": 1234567, "value": 5.0}
]
```

And it's a good point that if the single-entry `log_metric`, `log_param`, `set_tag` AbstractStore APIs are only used by the client/OSS server, we might as well eliminate them entirely in favor of log_batch - I created #985 to discuss that change, I'd like to get opinions from some other committers too before we decide to move forward with it - thanks!
